wasmplugin allows 'nil' value sha256

YonkaFang opened the pull request:
**Please provide a description of this PR:**

We used to allow the `Sha256` field to be optional, in fact istio sets `"nil"` to the null sha256 value to bypass the xDS API checksum and reverts this special value to `""` in the pilot-agent, see https://github.com/istio/istio/blob/ d14105866e811a925414e901b6386b3fbad4f800/pkg/wasm/convert.go#L217 .

We later changed the field to required in [Update WasmPlugin code to work with non-optional sha256 #35608](https://github.com/istio/istio/pull/35608), but kept the pilot-agent but kept the compatible logic in the pilot-agent.

I think it makes sense to allow not setting valid sha256 values in scenarios like testing, so I submitted this PR: The special value was released in the istio validation logic

istio-testing commented:
@YonkaFang: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 29fc09f765a9b8e98a048a86b541353d8f8f17f7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/45847/release-notes_istio/1676876982802976768) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

YonkaFang commented:
/test release-notes

YonkaFang commented:
> Why set `"nil"` instead of just `""`?

@howardjohn 

Let me explain the origin of this matter.

1. Firstly, it's because envoy (go-control-plane) validates this field and requires it to be non-empty:

   ```go
    if utf8.RuneCountInString(m.GetSha256()) < 1 {
        return RemoteDataSourceValidationError{
            field:  "Sha256",
            reason: "value length must be at least 1 runes",
        }
    }
   ```

2. Prior to https://github.com/istio/istio/pull/35608, the design in istio-api allowed sha256 to be empty. In order to pass the validation in go-control-plane, the Istio code would actually convert the empty value to the "nil" magic value and then convert it back in pilot-agent. However, https://github.com/istio/istio/pull/35608 may have considered this operation to be dirty and removed it. Therefore, even though Istio's validation logic allows this field to be empty, if the user actually sets it to empty, it will fail the validation in go-control-plane.

3. To solve the new conflict introduced in 2, and to support some lenient scenarios (such as testing) where a valid sha256 may not be necessary, I introduced the changes in this PR: users can choose to set sha256 to "nil" to achieve the effect of "ignoring sha256 validation".

howardjohn commented:
I still don't get it. There is Istio API and envoy API.

Why would we ever need the Istio API to be "nil" when we can just make it
empty?

Then we can translate that to envoy however - that seems irrelevant?

On Thu, Jul 6, 2023, 7:11 PM YonkaFang ***@***.***> wrote:

> Why set "nil" instead of just ""?
>
> @howardjohn <https://github.com/howardjohn>
>
> Let me explain the origin of this matter.
>
>    1.
>
>    Firstly, it's because envoy (go-control-plane) validates this field
>    and requires it to be non-empty:
>
>     if utf8.RuneCountInString(m.GetSha256()) < 1 {
>         return RemoteDataSourceValidationError{
>             field:  "Sha256",
>             reason: "value length must be at least 1 runes",
>         }
>     }
>
>    2.
>
>    Prior to #35608 <https://github.com/istio/istio/pull/35608>, the
>    design in istio-api allowed sha256 to be empty. In order to pass the
>    validation in go-control-plane, the Istio code would actually convert the
>    empty value to the "nil" magic value and then convert it back in
>    pilot-agent. However, #35608
>    <https://github.com/istio/istio/pull/35608> may have considered this
>    operation to be dirty and removed it. Therefore, even though Istio's
>    validation logic allows this field to be empty, if the user actually sets
>    it to empty, it will fail the validation in go-control-plane.
>    3.
>
>    To solve the new conflict introduced in 2, and to support some lenient
>    scenarios (such as testing) where a valid sha256 may not be necessary, I
>    introduced the changes in this PR: users can choose to set sha256 to "nil"
>    to achieve the effect of "ignoring sha256 validation".
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/45847#issuecomment-1624542367>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXLAJVDEOYENNQGS75LXO5V4RANCNFSM6AAAAAA2ADP7GQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

howardjohn commented:
wait.. are we converting Istio API to envoy API but then sending it to
Istio agent instead of envoy... but still using envoy validation?

That is.. very odd

On Thu, Jul 6, 2023, 7:16 PM John Howard ***@***.***> wrote:

> I still don't get it. There is Istio API and envoy API.
>
> Why would we ever need the Istio API to be "nil" when we can just make it
> empty?
>
> Then we can translate that to envoy however - that seems irrelevant?
>
> On Thu, Jul 6, 2023, 7:11 PM YonkaFang ***@***.***> wrote:
>
>> Why set "nil" instead of just ""?
>>
>> @howardjohn <https://github.com/howardjohn>
>>
>> Let me explain the origin of this matter.
>>
>>    1.
>>
>>    Firstly, it's because envoy (go-control-plane) validates this field
>>    and requires it to be non-empty:
>>
>>     if utf8.RuneCountInString(m.GetSha256()) < 1 {
>>         return RemoteDataSourceValidationError{
>>             field:  "Sha256",
>>             reason: "value length must be at least 1 runes",
>>         }
>>     }
>>
>>    2.
>>
>>    Prior to #35608 <https://github.com/istio/istio/pull/35608>, the
>>    design in istio-api allowed sha256 to be empty. In order to pass the
>>    validation in go-control-plane, the Istio code would actually convert the
>>    empty value to the "nil" magic value and then convert it back in
>>    pilot-agent. However, #35608
>>    <https://github.com/istio/istio/pull/35608> may have considered this
>>    operation to be dirty and removed it. Therefore, even though Istio's
>>    validation logic allows this field to be empty, if the user actually sets
>>    it to empty, it will fail the validation in go-control-plane.
>>    3.
>>
>>    To solve the new conflict introduced in 2, and to support some
>>    lenient scenarios (such as testing) where a valid sha256 may not be
>>    necessary, I introduced the changes in this PR: users can choose to set
>>    sha256 to "nil" to achieve the effect of "ignoring sha256 validation".
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/istio/istio/pull/45847#issuecomment-1624542367>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAEYGXLAJVDEOYENNQGS75LXO5V4RANCNFSM6AAAAAA2ADP7GQ>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

YonkaFang commented:
> 

@howardjohn 

Let me briefly describe it:

1. Envoy does not allow it to be empty;

2. Istio wants a more lenient restriction and has already implemented it (see [pkg/wasm/convert.go#L217](https://github.com/istio/istio/blob/d141058/pkg/wasm/convert.go#L217)), but ignored it in the Istio API validation;

3. Pilot-agent converts remote code to local code, and the xDS data actually sent to Envoy no longer includes sha256, so it is not limited by the validation logic in go-control-plane, as seen in [pkg/wasm/convert.go#L241C3-L241C3](https://github.com/istio/istio/blob/67bd6fc66a4a68089fd164a3f63fda6dc7cae547/pkg/wasm/convert.go#L241C3-L241C3):

```go
    // Rewrite remote fetch to local file.
    vm.Code = &core.AsyncDataSource{
        Specifier: &core.AsyncDataSource_Local{
            Local: &core.DataSource{
                Specifier: &core.DataSource_Filename{
                    Filename: f,
                },
            },
        },
    }
```

howardjohn commented:
> > 
> 
> @howardjohn
> 
> Let me briefly describe it:
> 
> 1. Envoy does not allow it to be empty;
> 2. Istio wants a more lenient restriction and has already implemented it (see [pkg/wasm/convert.go#L217](https://github.com/istio/istio/blob/d141058/pkg/wasm/convert.go#L217)), but ignored it in the Istio API validation;
> 3. Pilot-agent converts remote code to local code, and the xDS data actually sent to Envoy no longer includes sha256, so it is not limited by the validation logic in go-control-plane, as seen in [pkg/wasm/convert.go#L241C3-L241C3](https://github.com/istio/istio/blob/67bd6fc66a4a68089fd164a3f63fda6dc7cae547/pkg/wasm/convert.go#L241C3-L241C3):
> 
> ```go
>     // Rewrite remote fetch to local file.
>     vm.Code = &core.AsyncDataSource{
>         Specifier: &core.AsyncDataSource_Local{
>             Local: &core.DataSource{
>                 Specifier: &core.DataSource_Filename{
>                     Filename: f,
>                 },
>             },
>         },
>     }
> ```

None of this suggests the _user facing Istio API_ needs to change.

Here is how it should work


1. User creates a WasmPlugin with sha256="". This already is supported
2. We send this to pilot-agent as a `RemoteDataSource`. What we set in Sha256 here is purely an implementation detail.
3. Agent resolves based on RemoteDataSource and rewrites it

IIUC the issue is (2) is that somewhere is calling `.Validate()` on it?

If so, the solution seems to be to convert `""` to `"nil"` here https://github.com/istio/istio/blob/82c6704f85127bf7347a7b9af657b918443205e2/pilot/pkg/model/extensions.go#L232 perhaps?

Regardless, the _user_ should _never_ set `"nil"`

YonkaFang commented:
> Validate

@howardjohn 

Yes, our go-control-plane will call Validate during marshal to disallow empty sha256 values.

Previously, Istio had done this by changing empty sha256 values set by users to "nil". However, this behavior was removed in PR https://github.com/istio/istio/pull/35608. Currently, the simplest way I see to allow users to temporarily not set a valid sha256 value is to allow "nil" values in Istio's validation logic.

howardjohn commented:
Under no circumstances should the API allow anything other than a real
sha256 (which maybe we should validate is a valid one?) or "". It should
never accept "nil". That is leaking internal hacks into the user facing API.

On Mon, Jul 10, 2023 at 4:25 AM YonkaFang ***@***.***> wrote:

> Validate
>
> @howardjohn <https://github.com/howardjohn>
>
> Yes, our go-control-plane will call Validate during marshal to disallow
> empty sha256 values.
>
> Previously, Istio had done this by changing empty sha256 values set by
> users to "nil". However, this behavior was removed in PR #35608
> <https://github.com/istio/istio/pull/35608>. Currently, the simplest way
> I see to allow users to temporarily not set a valid sha256 value is to
> allow "nil" values in Istio's validation logic.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/45847#issuecomment-1628773029>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXJBQFXEW7PINNLCRADXPPRD7ANCNFSM6AAAAAA2ADP7GQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

YonkaFang commented:
> That

@howardjohn What do you think about retrieving the previous logic: change the "empty value" sha256 of istio-api allowed by Istio internally to "nil", which is semantically consistent and not exposed to users?

ingwonsong commented:
@YonkaFang Can you describe what you need to do?

As @howardjohn mentioned, use of empty `sha256` is totally implementation specific.
In other words, ECDS from Istiod is targeted to be passed to Istio Agent, but not Envoy directly.
So, I don't think that use of Envoy validation logic is appropriate to validate ECDS which comes from Istiod.

FYI, there are also the other fields in ECDS which is optional in Istio API surface, but not in Envoy API surface.

YonkaFang commented:
> 

@ingwonsong 

This is not a suitable question, but a current fact: the istio API allows sha256 to be empty, but in reality it will be flagged as an error by the validation code in go-control-plane when istio performs content serialization, so it is actually not usable.

What I want to do:

1. Make the "istio API allowing sha256 to be empty" work.

2. Reason for 1: In some testing scenarios, requiring a valid sha256 input can be cumbersome.

ingwonsong commented:
> go-control-plane will call Validate during marshal to disallow empty sha256 values.

@YonkaFang Can you navigate me to the code path you mentioned in the statement above?
